ipfsconn/ipfshttp/ipfshttp.go

```diff
@@ -288,57 +306,179 @@ func (ipfs *Connector) Pin(ctx context.Context, hash cid.Cid, maxDepth int) erro
+	ctx, cancelRequest := context.WithCancel(ctx)
+	defer cancelRequest()
+
+	switch ipfs.config.PinMethod {
+	case "refs":
+		// do refs -r first and timeout if we don't get at least
+		// one ref per pin timeout
+		outRefs := make(chan string)
+		go func() {
+			lastRefTime := time.Now()
+			ticker := time.NewTicker(progressTick)
+			defer ticker.Stop()
+			for {
+				select {
+				case <-ticker.C:
+					if time.Since(lastRefTime) >= ipfs.config.PinTimeout {
+						cancelRequest() // timeout
+						return
+					}
+				case <-outRefs:
+					lastRefTime = time.Now()
+				case <-ctx.Done():
+					return
+				}
+			}
+		}()
+
+		err := ipfs.refsProgress(ctx, hash, maxDepth, outRefs)
+		if err != nil {
+			return err
+		}
+
+		logger.Debugf("Refs for %s sucessfully fetched", hash)
+
 	}
 
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
+	// Pin request and timeout if there is no progress
+	outPins := make(chan int)
+	go func() {
+		var lastProgress int
+		lastProgressTime := time.Now()
 
-	timer := time.NewTimer(ipfs.config.PinTimeout)
-	defer timer.Stop()
+		ticker := time.NewTicker(ipfs.config.PinTimeout)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ticker.C:
+				if time.Since(lastProgressTime) > ipfs.config.PinTimeout {
+					// timeout request
+					cancelRequest()
+					return
+				}
+			case p := <-outPins:
+				// ipfs will send status messages every second
+				// or so but we need make sure there was
+				// progress by looking at number of nodes
+				// fetched.
+				if p > lastProgress {
+					lastProgress = p
+					lastProgressTime = time.Now()
+				}
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
 
-	reset := make(chan struct{})
-	defer close(reset)
+	err = ipfs.pinProgress(ctx, hash, maxDepth, outPins)
+	if err != nil {
+		return err
+	}
 
-	go checkTimeout(ctx, cancel, timer, ipfs.config.PinTimeout, reset)
+	logger.Info("IPFS Pin request succeeded: ", hash)
+	stats.Record(ctx, observations.Pins.M(1))
+	return nil
+}
 
-	switch ipfs.config.PinMethod {
-	case "refs": // do refs -r first
-		path := fmt.Sprintf("refs?arg=%s&%s", hash, pinArgs)
-		res, err := ipfs.doPostCtx(ctx, ipfs.client, ipfs.apiURL(), path, "", nil)
-		if err != nil {
-			return err
+// refsProgress fetches refs and puts them on a channel. Blocks until done or
+// error. refsProgress will always close the out channel. refsProgres will
+// not block on sending to the channel if it is full.
+func (ipfs *Connector) refsProgress(ctx context.Context, hash cid.Cid, maxDepth int, out chan<- string) error {
+	defer close(out)
+
+	ctx, span := trace.StartSpan(ctx, "ipfsconn/ipfshttp/refsProgress")
+	defer span.End()
+
+	path := fmt.Sprintf("refs?arg=%s&%s", hash, pinArgs(maxDepth))
+	res, err := ipfs.doPostCtx(ctx, ipfs.client, ipfs.apiURL(), path, "", nil)
+	if err != nil {
+		return err
+	}
+	defer res.Body.Close()
+
+	_, err = checkResponse(path, res)
+	if err != nil {
+		return err
+	}
```

Comment:
`checkResponse` should be removed

Reply:
https://github.com/ipfs/ipfs-cluster/pull/738/commits/c3ca53080c03ca7846cb6d50de9a040cd7bdc4d5

Reply:
Why would you do that? Please revert

Reply:
checkResponse reads the entire body with `readAll`(which reads till the stream ends) and throws error if status not `OK`. Why do we need this when we are reading the body using jsonDecoder?

Reply:
`checkResponse` reads an IPFS error message body when the request status is not 200. This is so you can read what the actual error message is (if any). If the body is empty the method will wrongly succeed.